From 3e9b947de830382c9ed71d838164fac63e0a26e0 Mon Sep 17 00:00:00 2001
From: "johan.bohlin" <johan.bohlin@intel.com>
Date: Fri, 28 Nov 2014 16:41:31 +0100
Subject: [PATCH] ASoC: Intel: BYTCR 5651 fix gpiod issue

Fix bug in gpio handling where gpio descriptor
was used after release.

Issue: GMINL-4445

Change-Id: I2673f60d98d096d9a03546cc0573f6a62fe24bfb
Signed-off-by: johan.bohlin <johan.bohlin@intel.com>
---
 sound/soc/intel/board/byt_cr_dpcm_rt5651.c | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/sound/soc/intel/board/byt_cr_dpcm_rt5651.c b/sound/soc/intel/board/byt_cr_dpcm_rt5651.c
index 8dbf187..04ecb47 100644
--- a/sound/soc/intel/board/byt_cr_dpcm_rt5651.c
+++ b/sound/soc/intel/board/byt_cr_dpcm_rt5651.c
@@ -582,13 +582,13 @@ static int byt_init(struct snd_soc_pcm_runtime *runtime)
 	desc = devm_gpiod_get_index(codec->dev, NULL, RT5651_GPIO_JD_INT);
 	if (!IS_ERR(desc)) {
 		drvdata->gpios.jd_int_gpio = desc_to_gpio(desc);
-		devm_gpiod_put(codec->dev, desc);
-
 		byt_export_gpio(desc, "JD-int");
 
 		pr_info("%s: GPIOs - JD-int: %d (pol = %d, val = %d)\n",
 			__func__, drvdata->gpios.jd_int_gpio,
 			gpiod_is_active_low(desc), gpiod_get_value(desc));
+
+		devm_gpiod_put(codec->dev, desc);
 	} else {
 		drvdata->gpios.jd_int_gpio = RT5651_GPIO_NA;
 		pr_err("%s: GPIOs - JD-int: Not present!\n", __func__);
@@ -597,13 +597,13 @@ static int byt_init(struct snd_soc_pcm_runtime *runtime)
 	desc = devm_gpiod_get_index(codec->dev, NULL, RT5651_GPIO_JD_INT2);
 	if (!IS_ERR(desc)) {
 		drvdata->gpios.jd_int2_gpio = desc_to_gpio(desc);
-		devm_gpiod_put(codec->dev, desc);
-
 		byt_export_gpio(desc, "JD-int2");
 
 		pr_info("%s: GPIOs - JD-int2: %d (pol = %d, val = %d)\n",
 			__func__, drvdata->gpios.jd_int2_gpio,
 			gpiod_is_active_low(desc), gpiod_get_value(desc));
+
+		devm_gpiod_put(codec->dev, desc);
 	} else {
 		drvdata->gpios.jd_int2_gpio = RT5651_GPIO_NA;
 		pr_warn("%s: GPIOs - JD-int2: Not present!\n", __func__);
@@ -612,8 +612,6 @@ static int byt_init(struct snd_soc_pcm_runtime *runtime)
 	desc = devm_gpiod_get_index(codec->dev, NULL, RT5651_GPIO_JACK_SWITCH);
 	if (!IS_ERR(desc)) {
 		drvdata->gpios.debug_mux_gpio = desc_to_gpio(desc);
-		devm_gpiod_put(codec->dev, desc);
-
 		byt_export_gpio(desc, "debug-mux");
 
 		dir = gpiod_get_direction(desc);
@@ -630,6 +628,8 @@ static int byt_init(struct snd_soc_pcm_runtime *runtime)
 		pr_info("%s: GPIOs - Debug-mux: %d (dir = %d, val = %d)\n",
 			__func__, drvdata->gpios.debug_mux_gpio, dir,
 			gpiod_get_value(desc));
+
+		devm_gpiod_put(codec->dev, desc);
 	} else {
 		drvdata->gpios.debug_mux_gpio = RT5651_GPIO_NA;
 		pr_warn("%s: GPIOs - Debug-mux: Not present!\n", __func__);
@@ -638,14 +638,14 @@ static int byt_init(struct snd_soc_pcm_runtime *runtime)
 	desc = devm_gpiod_get_index(codec->dev, NULL, RT5651_GPIO_ALC105_RESET);
 	if (!IS_ERR(desc)) {
 		drvdata->gpios.alc105_reset_gpio = desc_to_gpio(desc);
-		devm_gpiod_put(codec->dev, desc);
-
 		byt_export_gpio(desc, "ALC105");
 
 		pr_info("%s: GPIOs - ALC105: %d (pol = %d, val = %d)\n",
 			__func__, drvdata->gpios.alc105_reset_gpio,
 			gpiod_is_active_low(desc), gpiod_get_value(desc));
 
+		devm_gpiod_put(codec->dev, desc);
+
 	} else {
 		drvdata->gpios.alc105_reset_gpio = RT5651_GPIO_NA;
 		pr_warn("%s: GPIOs - ALC105 reset: Not present!\n", __func__);
@@ -654,14 +654,14 @@ static int byt_init(struct snd_soc_pcm_runtime *runtime)
 	desc = devm_gpiod_get_index(codec->dev, NULL, RT5651_GPIO_JD_BUTTONS);
 	if (!IS_ERR(desc)) {
 		drvdata->gpios.jd_buttons_gpio = desc_to_gpio(desc);
-		devm_gpiod_put(codec->dev, desc);
-
 		byt_export_gpio(desc, "JD-buttons");
 
 		pr_info("%s: GPIOs - JD-buttons: %d (pol = %d, val = %d)\n",
 			__func__, drvdata->gpios.jd_buttons_gpio,
 			gpiod_is_active_low(desc), gpiod_get_value(desc));
 
+		devm_gpiod_put(codec->dev, desc);
+
 	} else {
 		drvdata->gpios.jd_buttons_gpio = RT5651_GPIO_NA;
 		pr_warn("%s: GPIOs - JD-buttons: Not present!\n", __func__);
-- 
1.9.1

